File: generator/src/main/java/ch/bailu/gtk/converter/CallbackTable.java
package ch.bailu.gtk.converter;

import java.util.HashMap;
import java.util.Map;

import ch.bailu.gtk.tag.CallbackTag;

public class CallbackTable {
    private final Map<String, Map<String, CallbackTag>> table = new HashMap<>();


    private final static CallbackTable INSTANCE = new CallbackTable();

    public static CallbackTable instance() {
        return INSTANCE;
    }

    private CallbackTable() {}

    public void add(String namespace, CallbackTag callbackTag) {
        getTable(namespace).put(callbackTag.getName(), callbackTag);
    }

    private Map<String, CallbackTag> getTable(String namespace) {
        Map<String, CallbackTag> result = table.get(namespace);

        if (result == null) {
            result = new HashMap<>();
            table.put(namespace, result);
        }
        return result;
    }

    public boolean contains(String namespace, String name) {
        return getTable(namespace).get(name) != null;
    }

    public CallbackTag get(String namespace, String name) {
        return getTable(namespace).get(name);
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/table/CallbackTable.kt
package ch.bailu.gtk.table

import ch.bailu.gtk.log.Logable
import ch.bailu.gtk.parser.tag.CallbackTag
import java.io.Writer

object CallbackTable : Logable {
    private val table: MutableMap<String, MutableMap<String, CallbackTag?>> = HashMap()

    fun add(namespace: String, callbackTag: CallbackTag) {
        getTable(namespace)[callbackTag.getName()] = callbackTag
    }

    private fun getTable(namespace: String): MutableMap<String, CallbackTag?> {
        var result = table[namespace]
        if (result == null) {
            result = HashMap()
            table[namespace] = result
        }
        return result
    }

    operator fun get(namespace: String, name: String): CallbackTag? {
        return getTable(namespace)[name]
    }

    override fun log(writer: Writer) {
        writer.write("# ${CallbackTable.javaClass.canonicalName}\n")

        table.onEach { namespace ->
            writer.write("\n## ${namespace.key}\n\n")

            writer.write("| Name                                     | Detailed\n")
            writer.write("|------------------------------------------|---------\n")
            namespace.value.forEach {
                writer.write(String.format("| %-40s | %s\n", it.key, it.value))
            }
        }
    }
}
